src/high-level/polynomial-solver.lisp

```diff
@@ -0,0 +1,112 @@
+;;;; src/polynomial-solver.lisp
+;;;;
+;;;; Author: Juan M. Bello-Rivas
+
+(in-package #:magicl)
+
+;;; This file implements a simple univariate polynomial solver that relies on
+;;; the eigendecomposition of the companion matrix of the given polynomial to
+;;; obtain all of its roots.
+
+;;; A root refinement mechanism via Newton's iteration is included. Note that
+;;; polynomials (a) whose leading coefficient is much smaller than the
+;;; remaining coefficients or (b) having roots with multiplicity higher than
+;;; one will be harder to solve accurately.
+
+;;; A univariate polynomial is represented as a one-dimensional array of
+;;; complex coefficients.
+
+(deftype polynomial-coefficients () '(simple-array (complex double-float) (*)))
+
+(defstruct (polynomial (:constructor %make-polynomial))
+  (coefficients nil :type polynomial-coefficients))
```

Comment:
nitpick: declared type incompatible with default value

Reply:
I have thought about this problem before (not just for this PR but in general). With `(declaim (optimize safety))` it is impossible to instantiate a structure with empty coefficients because the compiler will signal a `type-error` whenever you call `(%make-polynomial)`. This check goes away when the safety setting is zero but, in my opinion, it is preferable to propagate the `polynomial-coefficients` type during compilation rather than letting SBCL emit code for `(or null polynomial-coefficients)` when it won't be `nil` in production. Nevertheless, I'm happy to change the type if the consensus among the team goes against my preference.

Reply:
I agree that changing the `:type` to `(or null foo-type)` is unsatisfactory if `NIL` isn't actually a valid value for the slot.  I usually make the default initform an `error` form in such cases to require the caller to provide a value, but I've certainly used `NIL` as well and don't feel that strongly about it.

I think it's actually implementation-dependent whether the `:type` is checked or enforced at all, but we heavily favor SBCL in other places, so it's not unreasonable to use its behavior as a reference when deciding such questions.